app/src/common/shared/com/igalia/wolvic/browser/WebAppsStore.java

```diff
@@ -0,0 +1,131 @@
+package com.igalia.wolvic.browser;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.os.Handler;
+import android.os.Looper;
+import android.preference.PreferenceManager;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.google.gson.Gson;
+import com.igalia.wolvic.R;
+import com.igalia.wolvic.ui.adapters.WebApp;
+import com.igalia.wolvic.utils.StringUtils;
+import com.igalia.wolvic.utils.SystemUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+public class WebAppsStore implements SharedPreferences.OnSharedPreferenceChangeListener {
+
+    protected final String LOGTAG = SystemUtils.createLogtag(this.getClass());
+
+    private Context mContext;
+    private List<WebApp> mWebApps;
+    private Set<WebAppsListener> mListeners;
+    private SharedPreferences mPrefs;
+
+    public WebAppsStore(Context context) {
+        mContext = context.getApplicationContext();
+        mWebApps = new ArrayList<>();
+        mListeners = new LinkedHashSet<>();
+
+        mPrefs = PreferenceManager.getDefaultSharedPreferences(mContext);
+        mPrefs.registerOnSharedPreferenceChangeListener(this);
+
+        mWebApps = new ArrayList<>();
+        updateWebAppsListFromStorage();
+    }
+
+    private void updateWebAppsListFromStorage() {
+        String json = SettingsStore.getInstance(mContext).getWebAppsData();
+
+        if (StringUtils.isEmpty(json)) {
+            mWebApps.clear();
+            notifyListeners();
+        } else {
+            try {
+                Gson gson = new Gson();
+                WebApp[] webAppsArray = gson.fromJson(json, WebApp[].class);
+                mWebApps.clear();
```

Comment:
This clear() is done in both branches so perhaps we could move it to the beginning of the method.